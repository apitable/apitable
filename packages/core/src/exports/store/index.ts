/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import * as ActionConstants from '../../modules/shared/store/action_constants';
import * as Reducers from './reducers';
import * as Selectors from './selectors';
import * as StoreActions from './actions';
export * from './interfaces';
export * from '../../modules/shared/store/constants';

export { ActionConstants, Reducers, Selectors, StoreActions };

