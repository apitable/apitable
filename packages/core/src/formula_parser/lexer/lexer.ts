/**
 * APITable <https://github.com/apitable/apitable>
 * Copyright (C) 2022 APITable Ltd. <https://apitable.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Token, TokenType } from './token';
import { t, Strings } from 'exports/i18n';
import { isNumber } from 'lodash';
export const EXPR_GRAMMAR: { key: TokenType; exp: RegExp }[] = [
  // The value in the record obtained by the field name constant
  { key: TokenType.Value, exp: /(\{\})|(\{(\\[{}])*[\s\S]*?[^\\]\})/ },
  // string literal
  { key: TokenType.String, exp: /(["”“](([^\\"“”])*?\\["”“])*.*?["”“])|(['‘’](([^\\'‘’])*?\\['‘’])*.*?['‘’])/ },
  // function name or field name constant
  { key: TokenType.Call, exp: /[^0-9.+\-|=*/><()（）!&%'"“”‘’^`~,，\s][^+\-|=*/><()（）!&%'"“”‘’^`~,，\s]*/ },
  // number literal
  { key: TokenType.Number, exp: /[0-9.]+/ },
  // not equal to
  { key: TokenType.NotEqual, exp: /!=/ },
  // and
  { key: TokenType.And, exp: /&&/ },
  // greater or equal to
  { key: TokenType.GreaterEqual, exp: />=/ },
  // less than or equal to
  { key: TokenType.LessEqual, exp: /<=/ },
  // or
  { key: TokenType.Or, exp: /\|\|/ },
  // comma, parameter separator
  { key: TokenType.Comma, exp: /[,，]/ },
  // Not
  { key: TokenType.Not, exp: /!/ },
  // add +
  { key: TokenType.Add, exp: /\+/ },
  // reduce -
  { key: TokenType.Minus, exp: /-/ },
  // take *
  { key: TokenType.Times, exp: /\*/ },
  // remove /
  { key: TokenType.Div, exp: /\// },
  // take remainder %
  { key: TokenType.Mod, exp: /%/ },
  // string concatenation
  { key: TokenType.Concat, exp: /&/ },
  // more than the
  { key: TokenType.Greater, exp: />/ },
  // less than
  { key: TokenType.Less, exp: /</ },
  // equal to
  { key: TokenType.Equal, exp: /=/ },
  // Left parenthesis
  { key: TokenType.LeftParen, exp: /[(（]/ },
  // closing parenthesis
  { key: TokenType.RightParen, exp: /[)）]/ },
  // whitespace characters
  { key: TokenType.Blank, exp: /\s+/ },
  // all other
  { key: TokenType.Unknown, exp: /.+/ },
];
// {} - {midterm test scores}
export interface ILexer {
  expression: string;
  currentTokenIndex: number;

  getNextToken(): Token | null;
  getPrevToken(): Token | null;
  reset(): void;
}
// The lexical analyzer of the formula, according to the incoming expression, exper -> token
export class FormulaExprLexer implements ILexer {
  readonly expression: string;

  readonly fullMatches: Token[];
  readonly matches: Token[];
  readonly errors: Error[] = []; // The lexer does not throw errors directly, but temporarily stores the errors for the caller to handle
  public currentIndex: number;

  constructor(expression: string) {
    this.expression = expression;

    this.fullMatches = this.getFullMatches();
    this.matches = this.filterUselessToken(this.fullMatches);
    this.currentIndex = -1;
  }

  get currentTokenIndex() {
    return this.currentIndex;
  }

  set currentTokenIndex(index: number) {
    if (!isNumber(index)) return;
    this.currentIndex = index;
  }

  getNextToken(): Token | null {
    this.currentIndex++;

    if (this.currentIndex > this.matches.length - 1) {
      return null;
    }

    const token = this.matches[this.currentIndex]!;
    return token;
  }

  getPrevToken(): Token | null {
    this.currentIndex--;

    if (this.currentIndex < 0) {
      return null;
    }

    const token = this.matches[this.currentIndex]!;
    return token;
  }

  reset(): void {
    this.currentIndex = -1;
  }

  private getFullMatches(): Token[] {
    const matched: string[] = this.expression.match(this.pattern()) || [];
    let index = 0;
    return matched.map((str, i) => {
      const token = this.tokenizer(index, str, matched[i + 1]);
      index += token.value.length;
      return token;
    });
  }

  private filterUselessToken(tokens: Token[]) {
    return tokens.filter(token => {
      if (token.type === TokenType.Unknown) {
        this.errors.push(
          new Error(
            t(Strings.function_err_unrecognized_operator, {
              token: token.value,
            }),
          ),
        );
      }
      return token.type !== TokenType.Blank && token.type !== TokenType.Unknown;
    });
  }

  private tokenizer(index: number, str: string, nextStr?: string): Token {
    for (let i = 0; i < EXPR_GRAMMAR.length; i++) {
      const key = EXPR_GRAMMAR[i]!.key;
      const regex = EXPR_GRAMMAR[i]!.exp;
      const type = key;

      if (regex.test(str)) {
        // When the tokenType matches Call, it is necessary to judge whether there is a left bracket, if not,
        // it means a pureValue without curly brackets
        if (type === TokenType.Call) {
          if (!nextStr || this.tokenizer(index, nextStr).type !== TokenType.LeftParen) {
            return new Token(TokenType.PureValue, index, str);
          }
        }
        return new Token(type, index, str);
      }
    }

    throw new Error(`Unexpected token: ${str}`);
  }

  private pattern(): RegExp {
    const pattern: string = EXPR_GRAMMAR.map(g => `(${g.exp.source})`).join('|');

    return new RegExp(pattern, 'g');
  }
}
