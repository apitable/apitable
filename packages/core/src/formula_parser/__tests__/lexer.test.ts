import { Strings, t } from 'exports/i18n';
import { FormulaExprLexer, ILexer, Token } from '../lexer';

describe('value', () => {
  it('should accept normal value', () => {
    const lexer = new FormulaExprLexer('{abc}    {🤯区间1  -a}  {   }');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept whitespaces', () => {
    const lexer = new FormulaExprLexer('{ Foo}{-　 }');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept escape sequence', () => {
    const lexer = new FormulaExprLexer('{}   {  x\\{_ _\\\\\\}}{}');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept empty name', () => {
    const lexer = new FormulaExprLexer('   {}  ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('string', () => {
  it('should accept empty strings', () => {
    const lexer = new FormulaExprLexer('""  "“  ”“ \'\'  \'’');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept Unicode quotation marks', () => {
    const lexer = new FormulaExprLexer('"abc"  "ab c“  ”ab c“');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept mismatched quotation marks', () => {
    const lexer = new FormulaExprLexer('"abc"  "ab c“  ”ab c"  ”ab c“');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept single-quotes', () => {
    const lexer = new FormulaExprLexer("   'abc'  'ab c’  ‘ab c’  ‘ab c'  ");
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept escape sequence', () => {
    const lexer = new FormulaExprLexer('   "ab \\"c"  ”a \\”c“  \' \\\'\' ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  test('escaping single-quotes in double-quoted strings is not allowed', () => {
    const lexer = new FormulaExprLexer("   \"ab \\'c\"  ”a \\’c“  ' \\'' ");
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  test('escaping double-quotes in single-quoted strings is not allowed', () => {
    const lexer = new FormulaExprLexer("   'ab \\\"c'  ’a \\”c‘  ' \\\"' ");
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('is converted into pure values if not closed', () => {
    const lexer = new FormulaExprLexer('   "ab c');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('accepts backslash as the last character in a string', () => {
    const lexer = new FormulaExprLexer('"ab c\\"');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('call', () => {
  it('should be followed by left parenthesis', () => {
    const lexer = new FormulaExprLexer('ab ab(');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept Unicode characters', () => {
    const lexer = new FormulaExprLexer('α👀()');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should be converted into a pure value if not immediately followed by a left parenthesis', () => {
    const lexer = new FormulaExprLexer('ab , c');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // TODO potential bug
  it('is converted into a pure value if the following left parenthesis is separated by spaces', () => {
    const lexer = new FormulaExprLexer('ab (');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('number', () => {
  it('should accept integers', () => {
    const lexer = new FormulaExprLexer('1244444444444444444444444444444444444444  78');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept trailing decimal point', () => {
    const lexer = new FormulaExprLexer('123. 0.');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept starting decimal point', () => {
    const lexer = new FormulaExprLexer('.123 .0');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('accepts single period', () => {
    const lexer = new FormulaExprLexer('.');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('accepts a sequence of periods', () => {
    const lexer = new FormulaExprLexer('123, ......');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('accepts multiple decimal points', () => {
    const lexer = new FormulaExprLexer('123......456');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('punctuation', () => {
  it('should accept operators with multiple characters', () => {
    const lexer = new FormulaExprLexer('!=&&>=<<==||');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accpet full-width commas', () => {
    const lexer = new FormulaExprLexer('， ,，,,');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should recognize single vertical bar as an unknown token', () => {
    const lexer = new FormulaExprLexer('| |');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept full-width parentheses', () => {
    const lexer = new FormulaExprLexer('（(）)');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('blank', () => {
  it('should recognize contiguous spaces as a single blank token', () => {
    const lexer = new FormulaExprLexer('12 3     A    ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should accept full-width spaces', () => {
    const lexer = new FormulaExprLexer('123 　　  A');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  it('should produce a single blank token if given a string with only spaces', () => {
    const lexer = new FormulaExprLexer('  　　  　');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('pure value', () => {
  it('should accept Unicode characters', () => {
    const lexer = new FormulaExprLexer('Foo a11 $T$ ᵣ☯');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('unknown token', () => {
  // FIXME potential bug
  it('is composed of only characters rejected by other tokens', () => {
    const lexer = new FormulaExprLexer('a~~~~~~~~~``~');
    expect(lexer.errors).toStrictEqual([new Error(t(Strings.function_err_unrecognized_operator, { token: '~~~~~~~~~``~' }))]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });

  // FIXME potential bug
  it('makes other tokens accept characters that should not be accepted', () => {
    const lexer = new FormulaExprLexer('abc ~~~ {}  ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toMatchSnapshot();
  });
});

describe('fullMatches', () => {
  it('should produce an empty token list if given an empty string', () => {
    const lexer = new FormulaExprLexer('');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.fullMatches).toStrictEqual([]);
  });
});

describe('matches', () => {
  it('should not contain blank tokens', () => {
    const lexer = new FormulaExprLexer('  A 123 "  "');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.matches).toMatchSnapshot();
  });

  it('should produce an empty token list if given an empty string', () => {
    const lexer = new FormulaExprLexer('');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.matches).toStrictEqual([]);
  });

  it('should produce an empty token list if given a string with only spaces', () => {
    const lexer = new FormulaExprLexer('        ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.matches).toStrictEqual([]);
  });

  it('should not contain unknown tokens', () => {
    const lexer = new FormulaExprLexer(' 123    ~~ ');
    expect(lexer.errors).toStrictEqual([]);
    expect(lexer.matches).toMatchSnapshot();
  });
});

describe('get token', () => {
  function collectTokens(lexer: ILexer) {
    const tokens: Token[] = [];
    let token: Token | null;
    while ((token = lexer.getNextToken())) {
      tokens.push(token);
    }
    return tokens;
  }

  describe('getNextToken', () => {
    it('should produce tokens in order', () => {
      const lexer = new FormulaExprLexer(' 123    ~~ ');
      expect(lexer.errors).toStrictEqual([]);
      const tokens = collectTokens(lexer);
      expect(tokens).toMatchSnapshot();
    });

    it('should produce null if no more token', () => {
      const lexer = new FormulaExprLexer('   ');
      expect(lexer.errors).toStrictEqual([]);
      expect(lexer.getNextToken()).toBeNull();
    });
  });

  describe('getPrevToken', () => {
    it('should produce previous token', () => {
      const lexer = new FormulaExprLexer(' 123 a  ');
      expect(lexer.errors).toStrictEqual([]);
      const token1 = lexer.getNextToken();
      const token2 = lexer.getNextToken();
      expect(lexer.getNextToken()).toBeNull();
      expect(lexer.getPrevToken()).toStrictEqual(token2);
      expect(lexer.getPrevToken()).toStrictEqual(token1);
    });

    it('should produce null if no previous token', () => {
      const lexer = new FormulaExprLexer(' 123  ');
      expect(lexer.errors).toStrictEqual([]);
      expect(lexer.getPrevToken()).toBeNull();
    });

    test('interleaving getNextToken and getPrevToken', () => {
      const lexer = new FormulaExprLexer(' 123 a=  ');
      expect(lexer.errors).toStrictEqual([]);
      const token1 = lexer.getNextToken();
      const token2 = lexer.getNextToken();
      expect(lexer.getPrevToken()).toStrictEqual(token1);
      expect(lexer.getPrevToken()).toBeNull();
      expect(lexer.getNextToken()).toStrictEqual(token1);
      expect(lexer.getNextToken()).toStrictEqual(token2);
      expect(lexer.getPrevToken()).toStrictEqual(token1);
      expect(lexer.getNextToken()).toStrictEqual(token2);
      const token3 = lexer.getNextToken();
      expect(lexer.getNextToken()).toBeNull();
      expect(lexer.getPrevToken()).toStrictEqual(token3);
    });
  });

  describe('reset', () => {
    it('should reset to the first token', () => {
      const lexer = new FormulaExprLexer(' 123 a=  ');
      expect(lexer.errors).toStrictEqual([]);
      const token1 = lexer.getNextToken();
      lexer.getNextToken();
      lexer.getNextToken();
      lexer.reset();
      expect(lexer.getNextToken()).toStrictEqual(token1);
    });
  });
});
