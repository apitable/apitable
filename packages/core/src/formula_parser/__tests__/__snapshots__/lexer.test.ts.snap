// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blank should accept full-width spaces 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "123",
  },
  Token {
    "index": 3,
    "type": "Blank",
    "value": " 　　  ",
  },
  Token {
    "index": 8,
    "type": "PureValue",
    "value": "A",
  },
]
`;

exports[`blank should produce a single blank token if given a string with only spaces 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "  　　  　",
  },
]
`;

exports[`blank should recognize contiguous spaces as a single blank token 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "12",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 3,
    "type": "Number",
    "value": "3",
  },
  Token {
    "index": 4,
    "type": "Blank",
    "value": "     ",
  },
  Token {
    "index": 9,
    "type": "PureValue",
    "value": "A",
  },
  Token {
    "index": 10,
    "type": "Blank",
    "value": "    ",
  },
]
`;

exports[`call is converted into a pure value if the following left parenthesis is separated by spaces 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "ab",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 3,
    "type": "LeftParen",
    "value": "(",
  },
]
`;

exports[`call should accept Unicode characters 1`] = `
[
  Token {
    "index": 0,
    "type": "Call",
    "value": "α👀",
  },
  Token {
    "index": 3,
    "type": "LeftParen",
    "value": "(",
  },
  Token {
    "index": 4,
    "type": "RightParen",
    "value": ")",
  },
]
`;

exports[`call should be converted into a pure value if not immediately followed by a left parenthesis 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "ab",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 3,
    "type": "Comma",
    "value": ",",
  },
  Token {
    "index": 4,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 5,
    "type": "PureValue",
    "value": "c",
  },
]
`;

exports[`call should be followed by left parenthesis 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "ab",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 3,
    "type": "Call",
    "value": "ab",
  },
  Token {
    "index": 5,
    "type": "LeftParen",
    "value": "(",
  },
]
`;

exports[`get token getNextToken should produce tokens in order 1`] = `
[
  Token {
    "index": 1,
    "type": "Number",
    "value": "123",
  },
]
`;

exports[`matches should not contain blank tokens 1`] = `
[
  Token {
    "index": 2,
    "type": "PureValue",
    "value": "A",
  },
  Token {
    "index": 4,
    "type": "Number",
    "value": "123",
  },
  Token {
    "index": 8,
    "type": "String",
    "value": ""  "",
  },
]
`;

exports[`matches should not contain unknown tokens 1`] = `
[
  Token {
    "index": 1,
    "type": "Number",
    "value": "123",
  },
]
`;

exports[`number accepts a sequence of periods 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "123",
  },
  Token {
    "index": 3,
    "type": "Comma",
    "value": ",",
  },
  Token {
    "index": 4,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 5,
    "type": "Number",
    "value": "......",
  },
]
`;

exports[`number accepts multiple decimal points 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "123......456",
  },
]
`;

exports[`number accepts single period 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": ".",
  },
]
`;

exports[`number should accept integers 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "1244444444444444444444444444444444444444",
  },
  Token {
    "index": 40,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 42,
    "type": "Number",
    "value": "78",
  },
]
`;

exports[`number should accept starting decimal point 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": ".123",
  },
  Token {
    "index": 4,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 5,
    "type": "Number",
    "value": ".0",
  },
]
`;

exports[`number should accept trailing decimal point 1`] = `
[
  Token {
    "index": 0,
    "type": "Number",
    "value": "123.",
  },
  Token {
    "index": 4,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 5,
    "type": "Number",
    "value": "0.",
  },
]
`;

exports[`punctuation should accept full-width parentheses 1`] = `
[
  Token {
    "index": 0,
    "type": "LeftParen",
    "value": "（",
  },
  Token {
    "index": 1,
    "type": "LeftParen",
    "value": "(",
  },
  Token {
    "index": 2,
    "type": "RightParen",
    "value": "）",
  },
  Token {
    "index": 3,
    "type": "RightParen",
    "value": ")",
  },
]
`;

exports[`punctuation should accept operators with multiple characters 1`] = `
[
  Token {
    "index": 0,
    "type": "NotEqual",
    "value": "!=",
  },
  Token {
    "index": 2,
    "type": "And",
    "value": "&&",
  },
  Token {
    "index": 4,
    "type": "GreaterEqual",
    "value": ">=",
  },
  Token {
    "index": 6,
    "type": "Less",
    "value": "<",
  },
  Token {
    "index": 7,
    "type": "LessEqual",
    "value": "<=",
  },
  Token {
    "index": 9,
    "type": "Equal",
    "value": "=",
  },
  Token {
    "index": 10,
    "type": "Or",
    "value": "||",
  },
]
`;

exports[`punctuation should accpet full-width commas 1`] = `
[
  Token {
    "index": 0,
    "type": "Comma",
    "value": "，",
  },
  Token {
    "index": 1,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 2,
    "type": "Comma",
    "value": ",",
  },
  Token {
    "index": 3,
    "type": "Comma",
    "value": "，",
  },
  Token {
    "index": 4,
    "type": "Comma",
    "value": ",",
  },
  Token {
    "index": 5,
    "type": "Comma",
    "value": ",",
  },
]
`;

exports[`punctuation should recognize single vertical bar as an unknown token 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "| |",
  },
]
`;

exports[`pure value should accept Unicode characters 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "Foo",
  },
  Token {
    "index": 3,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 4,
    "type": "PureValue",
    "value": "a11",
  },
  Token {
    "index": 7,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 8,
    "type": "PureValue",
    "value": "$T$",
  },
  Token {
    "index": 11,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 12,
    "type": "PureValue",
    "value": "ᵣ☯",
  },
]
`;

exports[`string accepts backslash as the last character in a string 1`] = `
[
  Token {
    "index": 0,
    "type": "String",
    "value": ""ab c\\"",
  },
]
`;

exports[`string escaping double-quotes in single-quoted strings is not allowed 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "String",
    "value": "'ab \\"c'",
  },
  Token {
    "index": 11,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 13,
    "type": "String",
    "value": "’a \\”c‘",
  },
  Token {
    "index": 20,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 22,
    "type": "String",
    "value": "' \\"'",
  },
  Token {
    "index": 27,
    "type": "Blank",
    "value": " ",
  },
]
`;

exports[`string escaping single-quotes in double-quoted strings is not allowed 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "String",
    "value": ""ab \\'c"",
  },
  Token {
    "index": 11,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 13,
    "type": "String",
    "value": "”a \\’c“",
  },
  Token {
    "index": 20,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 22,
    "type": "String",
    "value": "' \\''",
  },
  Token {
    "index": 27,
    "type": "Blank",
    "value": " ",
  },
]
`;

exports[`string is converted into pure values if not closed 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "PureValue",
    "value": ""ab c",
  },
]
`;

exports[`string should accept Unicode quotation marks 1`] = `
[
  Token {
    "index": 0,
    "type": "String",
    "value": ""abc"",
  },
  Token {
    "index": 5,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 7,
    "type": "String",
    "value": ""ab c“",
  },
  Token {
    "index": 13,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 15,
    "type": "String",
    "value": "”ab c“",
  },
]
`;

exports[`string should accept empty strings 1`] = `
[
  Token {
    "index": 0,
    "type": "String",
    "value": """",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 4,
    "type": "String",
    "value": ""“",
  },
  Token {
    "index": 6,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 8,
    "type": "String",
    "value": "”“",
  },
  Token {
    "index": 10,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 11,
    "type": "String",
    "value": "''",
  },
  Token {
    "index": 13,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 15,
    "type": "String",
    "value": "'’",
  },
]
`;

exports[`string should accept escape sequence 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "String",
    "value": ""ab \\"c"",
  },
  Token {
    "index": 11,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 13,
    "type": "String",
    "value": "”a \\”c“",
  },
  Token {
    "index": 20,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 22,
    "type": "String",
    "value": "' \\''",
  },
  Token {
    "index": 27,
    "type": "Blank",
    "value": " ",
  },
]
`;

exports[`string should accept mismatched quotation marks 1`] = `
[
  Token {
    "index": 0,
    "type": "String",
    "value": ""abc"",
  },
  Token {
    "index": 5,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 7,
    "type": "String",
    "value": ""ab c“",
  },
  Token {
    "index": 13,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 15,
    "type": "String",
    "value": "”ab c"",
  },
  Token {
    "index": 21,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 23,
    "type": "String",
    "value": "”ab c“",
  },
]
`;

exports[`string should accept single-quotes 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "String",
    "value": "'abc'",
  },
  Token {
    "index": 8,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 10,
    "type": "String",
    "value": "'ab c’",
  },
  Token {
    "index": 16,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 18,
    "type": "String",
    "value": "‘ab c’",
  },
  Token {
    "index": 24,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 26,
    "type": "String",
    "value": "‘ab c'",
  },
  Token {
    "index": 32,
    "type": "Blank",
    "value": "  ",
  },
]
`;

exports[`unknown token is composed of only characters rejected by other tokens 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "a",
  },
  Token {
    "index": 1,
    "type": "Unknown",
    "value": "~~~~~~~~~\`\`~",
  },
]
`;

exports[`unknown token makes other tokens accept characters that should not be accepted 1`] = `
[
  Token {
    "index": 0,
    "type": "PureValue",
    "value": "abc",
  },
  Token {
    "index": 3,
    "type": "Blank",
    "value": " ",
  },
  Token {
    "index": 4,
    "type": "Value",
    "value": "~~~ {}  ",
  },
]
`;

exports[`value should accept empty name 1`] = `
[
  Token {
    "index": 0,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 3,
    "type": "Value",
    "value": "{}",
  },
  Token {
    "index": 5,
    "type": "Blank",
    "value": "  ",
  },
]
`;

exports[`value should accept escape sequence 1`] = `
[
  Token {
    "index": 0,
    "type": "Value",
    "value": "{}",
  },
  Token {
    "index": 2,
    "type": "Blank",
    "value": "   ",
  },
  Token {
    "index": 5,
    "type": "Value",
    "value": "{  x\\{_ _\\\\\\}}",
  },
  Token {
    "index": 19,
    "type": "Value",
    "value": "{}",
  },
]
`;

exports[`value should accept normal value 1`] = `
[
  Token {
    "index": 0,
    "type": "Value",
    "value": "{abc}",
  },
  Token {
    "index": 5,
    "type": "Blank",
    "value": "    ",
  },
  Token {
    "index": 9,
    "type": "Value",
    "value": "{🤯区间1  -a}",
  },
  Token {
    "index": 20,
    "type": "Blank",
    "value": "  ",
  },
  Token {
    "index": 22,
    "type": "Value",
    "value": "{   }",
  },
]
`;

exports[`value should accept whitespaces 1`] = `
[
  Token {
    "index": 0,
    "type": "Value",
    "value": "{ Foo}",
  },
  Token {
    "index": 6,
    "type": "Value",
    "value": "{-　 }",
  },
]
`;
